src/Validation.ts

```diff
@@ -0,0 +1,215 @@
+type Fork = (
+    reject: (value: Error) => void,
+    resolve: (value: Valid | Invalid) => void
+) => void;
+
+export interface InvalidResult {
+    value: any;
+    message: string;
+    key?: string[];
+}
+
+export interface Valid {
+    isValid: true;
+    x: undefined;
+    fork: undefined;
+    and(
+        other: Valid | Invalid | AsyncValidation
+    ): Valid | Invalid | AsyncValidation;
+    and(other: Valid): Valid;
+    and(other: Invalid): Invalid;
+    and(other: AsyncValidation): AsyncValidation;
+    or(other: Valid | Invalid | AsyncValidation): Valid | AsyncValidation;
+    or(other: Valid | Invalid): Valid;
+    or(other: AsyncValidation): AsyncValidation;
+    format: (
+        fn: (
+            invalidResult: InvalidResult,
+            index: number,
+            all: InvalidResult[]
+        ) => InvalidResult
+    ) => Valid;
+    getResult: () => void;
+}
+
+export interface Invalid {
```

Comment:
This type name is a bit mysterious. Invalid what?

Reply:
It's an Invalid Validation. It represents an Invalid validation result.
It is also part of the monoid Validation.
A monoid is a set of value that can be combined together by an operation into other member of the same set.
Here the set of Validation is comprised of Valid, Invalid and Async.
IMO Invalid is the right name.